### ISSUE TITLE
reportUnusedExpression false positive when used with logical operators

### ISSUE THREAD

**Issue by kennipj** [microsoft/pyright#3550]
Consider the following example:

```py
def some_func(should_print: bool) -> None:
    should_print and print("something") # warning: Expression value is unused
```

Due to short-circuiting logic the print statement is only executed if `should_print` evaluates to `True`, and hence the expression value `should_print` is not unused as the warning would suggest.

The same warning also appears when used with the `or` operator.

Tested on pyright 1.1.252

**Comment by erictraut** [comment]
That's an inadvisable use of short-circuiting. Why wouldn't you write the code the normal (more readable) way instead?

```python
if should_print:
    print("something")
```

The `reportUnusedExpression` check has known potential for false positives, which is one of the reasons it emits an error only in strict mode. If want to benefit from the check, you should write your code to work with it.

**Comment by kennipj** [comment]
> That's an inadvisable use of short-circuiting. Why wouldn't you write the code the normal (more readable) way instead?

I agree, an if statement is a lot clearer. I came across this code by chance, and figured I may as well report it :)

**Comment by Ohhama1** [comment]
mul_func は無反応code ではなく、メッセージは表示できるはずです。
